app/decorators/provider_decorator.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class ProviderDecorator < BaseDecorator
+  def name_and_role
+    return t('generic.not_available') if name_and_role_blank?
+
+    "#{formatted_name} (#{formatted_role})"
+  end
+
+  private
+
+  def name_and_role_blank?
+    [(respond_to?(:name) && name.blank?),
+     (respond_to?(:role) && role.blank?)].all?
```

Comment:
Is it possible for a provider not to have either of these attributes? From the method you are replacing in lib/court_data_adaptor/resource/provider.rb it looks as though these attributes are presumed to exist. In this case, this check could simply be `name.blank? && role.blank?`.

Reply:
> Is it possible for a provider not to have either of these attributes? From the method you are replacing in lib/court_data_adaptor/resource/provider.rb it looks as though these attributes are presumed to exist. In this case, this check could simply be `name.blank? && role.blank?`.

I spent a fair bit of time mucking about with this. To get the desired result, which is defensive because of the lack of knowledge of what we can expect from CP, I have used this combination. 

Take a look at the tests for it

Reply:
(Please bear with me as I try to understand how it works)

In the `CourtDataAdapter::Resource::Provider` the `name` and `role` are explicitly defined here:

https://github.com/ministryofjustice/laa-court-data-ui/blob/d61abdc6f88778f2c87e4fdc25970f7fcbaa665d/lib/court_data_adaptor/resource/provider.rb#L10-L11

So those attributes will always exist, even if they are not returned from the CP, but they may be `nil`.

As it says in the [Json Api Client documentation](https://github.com/JsonApiClient/json_api_client#schema) these definitions are optional but when they are present it handles default values and type casting.

Reply:
Yes,  because they are explicitly defined in the resource object these attributes/methods will always be available and be at least nil and therefore the `respond_to` is not needed.

However, these [tests were done in isolation](https://github.com/ministryofjustice/laa-court-data-ui/pull/394/files#diff-16f7eab620be5bf8db87c7798b3b0fab45ab6690e54ae8bb1c3ae6f26c45b222R52-R62) as part of presenter logic alone and so were made to handle the possibility that they may raise an error. 

The tests and therefore the checks may be excessive but it is intended to eventually move the court data adaptor lib to a gem, in which case the presenter logic is all we have. I wonder if we may need to be more careful with other attributes too.

For simplicity I could remove these tests and checks for now?!

Reply:
I can see that, when moving the adaptor to a gem, the checks that `name` and `role` exist will be separated from the presenter. However, isn't it the purpose of the adaptor to ensure that a consistent public interface is presented regardless of what appears upstream? I suppose the question is whether it is likely that these attributes will be removed from the gems interface in the future. If this happens I would expect them to report a deprecation warning in (at least) one version, rather than just being removed.

Reply:
I will remove for now as is obviously controversial and excessive in any event

Reply:
done now